server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java

```diff
@@ -0,0 +1,140 @@
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.VersionedNamedWriteable;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * {@code RankContextBuilder} is used as a base class to manage input, parsing,
+ * and subsequent generation of appropriate contexts for handling searches that
+ * require multiple queries for global rank relevance.
+ *
+ * This class contains a {@code List<QueryBuilder>}s, size, and from members that
+ * sub-classes use to execute queries to generate ranking results. These are not part
+ * of ranking input, but instead generated from their respective values within a search
+ * query. We cannot generate the list of queries until after any kNN queries are executed.
+ */
+public abstract class RankBuilder<RCB extends RankBuilder<RCB>> implements VersionedNamedWriteable, ToXContentObject {
+
+    public static final ParseField WINDOW_SIZE_FIELD = new ParseField("window_size");
+
+    public static final int DEFAULT_WINDOW_SIZE = SearchService.DEFAULT_SIZE;
+
+    protected int windowSize = DEFAULT_WINDOW_SIZE;
+
+    public RankBuilder() {}
+
+    public RankBuilder(StreamInput in) throws IOException {
+        windowSize = in.readVInt();
+    }
+
+    public final void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(windowSize);
+        doWriteTo(out);
+    }
+
+    public abstract void doWriteTo(StreamOutput out) throws IOException;
+
+    @Override
+    public final XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.startObject(getWriteableName());
+        builder.field(WINDOW_SIZE_FIELD.getPreferredName(), windowSize);
+        doXContent(builder, params);
+        builder.endObject();
+        builder.endObject();
+        return builder;
+    }
+
+    protected abstract void doXContent(XContentBuilder builder, Params params) throws IOException;
+
+    /**
+     * Allows additional validation as part of {@link SearchRequest#validate} based on subclass rank parameters.
+     */
+    public abstract ActionRequestValidationException validate(
+        ActionRequestValidationException validationException,
+        SearchSourceBuilder source
+    );
```

Comment:
I moved this exception to RankBuilder for the minimum query count, but I think it still fits better to have the rank constant be a validation exception since it should be caught as part of the original request. To accommodate this, I've added a doValidate in a similar fashion to doEquals and doHashCode.

Reply:
Thanks. Validate is convenient, but ideally you can fail fast as soon as you set something that is wrong, and you see its validation code directly in the setter. Validate is still useful for places where you need to check more than one value and you need the object to be fully initialized.